js/src/ipc/message.ts

```diff
@@ -104,6 +107,9 @@ export class AsyncMessageReader implements AsyncIterableIterator<Message> {
     public async next(): Promise<IteratorResult<Message>> {
         let r;
         if ((r = await this.readMetadataLength()).done) { return ITERATOR_DONE; }
+        // ARROW-6313: If the first 4 bytes are continuation message, read the next 4 for the length
+        if ((r.value === -1) &&
+            (r = await this.readMetadataLength()).done) { return ITERATOR_DONE; }
```

Comment:
It's too bad we have to duplicate this code. Could the readers be combined somehow? Not in this PR, just curious

Reply:
Unfortunately not. Async functions are implemented via promises, which don’t allow parameterizing the scheduler.  Async functions force the caller to handle the returned promise, guaranteeing the function doesn’t finish its work synchronously, even if it could. 

This is a key difference between JS promises and Python’s asyncio module, which does execute synchronously if no actual async work is performed.